src/instrumenter/transpile.ts

```diff
@@ -406,21 +509,41 @@ function transpileFunctionCall(expr: SFunctionCall, ctx: TranspilingContext): Ex
         expr.callee.defSite === "builtin_fun"
     ) {
         const arg = single(expr.args);
-        const [sVar, path] = decomposeStateVarRef(arg);
+        const argT = ctx.typeEnv.typeOf(arg);
+
+        assert(argT instanceof PointerType, `sum expects a pointer to array/map, not ${argT.pp()}`);
+
+        if (argT.to instanceof MappingType) {
+            const [sVar, path] = decomposeStateVarRef(arg);
+            assert(
+                sVar !== undefined,
+                `sum argument should be a state var(or a part of it), not ${arg.pp()}`
+            );
+
+            const lib = ctx.instrCtx.sVarToLibraryMap.get(
+                sVar,
+                path.map((el) => (el instanceof SNode ? null : el))
+            );
+
+            assert(lib !== undefined, `State var ${sVar.name} should already be interposed`);
+
+            const struct = single(lib.vStructs);
+            return mkStructFieldAcc(factory, transpile(arg, ctx), struct, "sum");
+        }
+
         assert(
-            sVar !== undefined,
-            `sum argument should be a state var(or a part of it), not ${arg.pp()}`
+            argT.to instanceof ArrayType,
+            `sum expects a pointer to array/map, not ${argT.pp()}`
```

Comment:
A note, that does not unrgently require a fix: `pp()`-ing in `assert()`s is slow, especially if used in message template string. This is because arguments are resolved on each call, so `pp()` is actually executed no matter if assertion is `true` or `false`. I guess we need `assert()` optimization in **solc-typed-ast**, or local optimization to boost performance.

Reply:
Yeah, lets do a separate cleanup pass for that. I have 2-3 things in mind we can do in a single cleanup pass. Lets discuss offline